tensorflow/core/ops/math_ops.cc

```diff
@@ -509,6 +509,83 @@ REGISTER_OP("TruncateDiv")
 REGISTER_OP("RealDiv").BINARY_MORE().SetShapeFn(
     shape_inference::BroadcastBinaryOpShapeFn);
 
+Status BroadcastFMAOpOutputShape(InferenceContext* c) {
+  ShapeHandle out;
+  out = c->input(0);
+  c->Merge(out, c->input(1), &out);
+  c->Merge(out, c->input(2), &out);
+  c->set_output(0, out);
+
+  return Status::OK();
+}
+
+Status BroadcastFMA2OpOutputShape(InferenceContext* c) {
+  ShapeHandle out;
+  /*
+  DimensionHandle d0, d1, d2;
+  d0 = c->NumElements(c->input(0));
+  d1 = c->NumElements(c->input(1));
+  d2 = c->NumElements(c->input(2));  
+  if(!c->ValueKnown(d0) || c->Value(d0)!=1)
+    out = c->input(0);
+  else if(!c->ValueKnown(d1) || c->Value(d1)!=1)
+    out = c->input(1);
+  else if(!c->ValueKnown(d2) || c->Value(d2)!=1)
+    out = c->input(2);
+  else
+    out = c->input(3);
+    */
+  out = c->input(0);
+  c->Merge(out, c->input(1), &out);
+  c->Merge(out, c->input(2), &out);
+  c->Merge(out, c->input(3), &out);
+
+  c->set_output(0, out);
+  return Status::OK();
+}
+
+REGISTER_OP("FusedMulAdd")
+    .Input("x1: T")
+    .Input("y1: T")
+    .Input("x2: T")
+    .Output("z: T")
+    .Attr("T: {bfloat16, half, float, double}")
+    .SetShapeFn(BroadcastFMAOpOutputShape);
+
+REGISTER_OP("FusedMulAdd2")
+    .Input("x1: T")
+    .Input("y1: T")
+    .Input("x2: T")
+    .Input("y2: T")
+    .Output("z: T")
+    .Attr("T: {bfloat16, half, float, double}")
+    .SetShapeFn(BroadcastFMA2OpOutputShape);
+
+REGISTER_OP("FusedMulSub")
+    .Input("x1: T")
+    .Input("y1: T")
+    .Input("x2: T")
+    .Output("z: T")
+    .Attr("T: {bfloat16, half, float, double}")
+    .SetShapeFn(BroadcastFMAOpOutputShape);
+
+REGISTER_OP("FusedMulSubRev")
+    .Input("x1: T")
+    .Input("y1: T")
+    .Input("x2: T")
+    .Output("z: T")
+    .Attr("T: {bfloat16, half, float, double}")
+    .SetShapeFn(BroadcastFMAOpOutputShape);
+
+REGISTER_OP("FusedMulSub2")
+    .Input("x1: T")
+    .Input("y1: T")
+    .Input("x2: T")
+    .Input("y2: T")
+    .Output("z: T")
+    .Attr("T: {bfloat16, half, float, double}")
+    .SetShapeFn(BroadcastFMA2OpOutputShape);
```

Comment:
This is a non-trivial interface change, and will probably needs approval from upstream. I'd suggest to also upstream this early should be decided to make this new API public.

Btw, this is going to break API compatibility test. I'd suggest to update golden files by compile and running api_compatibility_test to allow it update the golden APIs
`//tensorflow/tools/api/tests:api_compatibility_test`. Supply the `--update_goldens True` flag to the test case to allow it update the interface files.

Reply:
I'm not familiar with the API compatibility test. Would it be satisfied if I added '_' to all names (_FusedMulAdd, etc.) to make them internal?

Reply:
To my understanding adding underscore in the naming will not make them internal. My understanding is that registration inside `tensorflow/core/ops` will make the API public. I think we can go three ways:

- Option 1: Continue down the path you have. We fix the API compatibility test together with functionality tests. Then we submit upstream and address any concerns they have. Heads up though, they may turn down this new API. If that happens, we can try to address their concerns from there.
- Option 2: Continue down the path until we fixed all functionality test. Before we upstream it, we remove all updated components in `tensorflow/core/ops` and `tensorflow/python/kernel_tests`. (We can instead add C++ grappler unit tests instead to make sure there is test coverage on it.) This ensures that the operator stays private and we are less likely turned down by upstream.
- Option 3:  Effectively go with option 1, if they turn us down, get approval from tech-lead/owner, and merge it to `master-rocm-enhanced`. I personally do not like this option because your code changes in `FusedMulAdd` is a generic change that does not depend on specific `ROCm` features.

Reply:
Is it possible to keep the ops available for grappler without registering them in tensorflow/core/ops?

Reply:
@ekuznetsov139 I haven't worked on grappler before so I don't know. I remember that @deven-amd used to work on a grappler feature. 

@deven-amd What did you do with the ops used in the grappler pass? Do you have to register it before you can use it?

Reply:
I think putting a underscore before the op name makes the op "internal"

The changes I made for the fusion work I did last year can be seen in this PR 
https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/782

I did indeed have an underscore prefix for all the op names!